### ISSUE TITLE
Allow KVC Collection Operators in Sort Descriptors

### ISSUE THREAD

**Issue by bdkjones** [realm/realm-swift#8666]
### Problem

Consider this Object:

```
final class Foo: Object
{
      @Persisted(originProperty: "parentFoos") var associatedThings: LinkingObjects<Baz>
}
```

Suppose I have a bunch of `Foo` objects that get displayed in an `NSTableView`. I would like to be able to sort these objects by `associatedThings.@count`. But I can't do that. The error I get is `Cannot sort on key path 'associatedThings.@count': KVC collection operators are not supported.`

This is frustrating because I have to add a new property to `Foo` that lets me sort them:

```
final class Foo: Object
{
      @Persisted(originProperty: "parentFoos") var associatedThings: LinkingObjects<Baz>
      @Persisted var associatedThingsCount: Int = 0
}
```

Then I have to keep this count property up-to-date manually and very carefully manage it. It's a big headache, a waste of database space, and overhead for the Atlas sync service.

### Solution

Realm should make it easy to sort `Object` subclasses by the count of how many objects are in a `List`, `MutableSet`, or `LinkingObjects`. 

### Alternatives

Hacky extra properties that don't serve any purpose other than to enable sorting. 

### How important is this improvement for you?

Would be a major improvement

### Feature would mainly be used with

Atlas Device Sync

**Comment by Jaycyn** [comment]
If I understand correctly, you have a Baz class with a forward relationship property that contains a Foo, and each Foo has a reverse relationship back to Baz. You want to read your Foo's and sort them by the count of Baz each one backlinks to

Maybe not the answer you wanted because it would be great to be able to use @count but this is at least a solution. Based on these models and assuming

foo0 backlinks to 3 baz
foo1 backlinks to 2 baz
foo2 backlinks to 1 baz

```
class Baz: Object {
    @Persisted var bazName = ""
    @Persisted var myFoo: Foo!
}

class Foo: Object {
    @Persisted var fooName = ""
    @Persisted(originProperty: "myFoo") var linkedThings: LinkingObjects<Baz>
}
```

This code will read the Foos and sort them ascending by the count of backlinked Baz in each:

```
let results = realm.objects(Foo.self).sorted {$0.linkedThings.count < $1.linkedThings.count }
for foo in results {
    print(foo.fooName, foo.linkedThings.count)
}
```

and the output

```
f2 1
f1 2
f0 3
```

**Comment by bdkjones** [comment]
@Jaycyn yep, I'm aware of how to do it in code. The problem (as my giant note indicated) is that I need to be able to construct an `NSSortDescriptor` to assign to an `NSTableColumn` so that the table sorts itself correctly when that column header is clicked.

I cannot currently do that for the count of objects in a Realm collection.

**Comment by Jaycyn** [comment]
Yup. Just providing a workaround for the sorting component (which you know) and +1 to add support for @count so it can be used in TableViews.

I do seem to recall a project where I ran into the same issue and worked around it with NSTableView:sortDescriptorsDidChange (I think), ignored the actual sortDescriptors and managed the dataSource separately from the tableView. Could be wrong on that as it's beed a few years.

**Comment by Jaycyn** [comment]
This may not be any help at all, or perhaps you MUST use NSSortDescriptor but but because there was no support for @count, I had put in a Realm feature request for that TWO years ago (hrrrrmmmfff) and ended up ditching the NSSortDescriptor entirely and powering the TableView separately.

The Item Object model had just an itemName and itemDescription property so this sorted by either.

```
    func tableView(_ tableView: NSTableView, sortDescriptorsDidChange oldDescriptors: [NSSortDescriptor]) {
        print("test")

        guard let descriptor = tableView.sortDescriptors.first else { return } //the sortDescriptor key is the realm objects property name for this col
        let asc = descriptor.ascending //toggles true to false
        let colToSortBy = descriptor.key! //the key is the actual realm object property name to sort by

        switch colToSortBy {
        case ColumnIdentifiers.itemName:
            if asc == true {
                self.realmData.sorted {$0.linkedThings.count < $1.linkedThings.count }
            } else {
                self.realmData.sorted {$0.linkedThings.count > $1.linkedThings.count }
            }
        case ColumnIdentifiers.itemDesc:
            if asc == true {
                self.realmData.sorted {$0.linkedOtherThings.count < $1.linkedOtherThings.count }
            } else {
                self.realmData.sorted {$0.linkedOtherThings.count > $1.linkedOtherThings.count }
            }
        default:
            break
        }

        self.itemTableView.reloadData()
    }
```

**Comment by bdkjones** [comment]
@Jaycyn That's a clever workaround, but my TableView has 33 columns!
